source/extensions/filters/network/redis_proxy/command_splitter_impl.cc

```diff
@@ -900,6 +901,293 @@ void MSETRequest::onChildResponse(Common::Redis::RespValuePtr&& value, uint32_t
+    std::string index = cursor.substr(cursor.length() - 4);
+    cursor = cursor.substr(0, cursor.length() - 4);
+    shard_idx = std::stoi(index);
+  }
+
+  // Completely reconstructing the request to add/modify count since we can't override the incoming array directly
+  // Add the command and cursor to the request array
+  addBulkString(requestArray, "SCAN");
+  addBulkString(requestArray, cursor);
+
+  // Iterate over the arguments modify count if necessary
+  // We are setting 1000 as the default count value
+  for (size_t i = 2; i < incoming_request->asArray().size(); ++i) {
+      std::string arg = incoming_request->asArray()[i].asString();
+      addBulkString(requestArray, arg);
+
+      // Check if the argument requires a value
+      if (requiresValue(arg)) {
+          if (arg == "count") {
+            addBulkString(requestArray, count);
+            ++i;
+          } else {
+            // If the current argument requires a value, add it to the request array
+            std::string value = incoming_request->asArray()[++i].asString();
+            addBulkString(requestArray, value);
+          }
+      }
+  }
+
+  // If the "count" argument is not found, add it with the default value
+  if (incoming_request->asArray().size() == 2) {
+      addBulkString(requestArray, "count");
+      addBulkString(requestArray, count);
+  }
+
+  std::unique_ptr<ScanRequest> request_ptr{
+      new ScanRequest(callbacks, command_stats, time_source, delay_command_latency)};
+
+  // caching the request and route for making child request from response
+  request_ptr->request_ = requestArray;
+  request_ptr->route_ = router.upstreamPool(key, stream_info);
+
+  if (request_ptr->route_) {
+    Extensions::NetworkFilters::RedisProxy::ConnPool::InstanceImpl* instance =
+        dynamic_cast<Extensions::NetworkFilters::RedisProxy::ConnPool::InstanceImpl*>(
+            request_ptr->route_->upstream(key).get());
+
+    request_ptr->num_of_Shards_ = instance->getNumofRedisShards();
+    if (request_ptr->num_of_Shards_ == 0 ) {
+      callbacks.onResponse(Common::Redis::Utility::makeError(Response::get().NoUpstreamHost));
+    }
+  }
+  else{
+    callbacks.onResponse(Common::Redis::Utility::makeError(Response::get().NoUpstreamHost));
+  }
+
+  // If shard index is some random value, we are setting the shard to 0 to avoid crashing
+  // This ensures scan always returns value.
+  if (shard_idx > request_ptr->num_of_Shards_) {
+    shard_idx = 0;
+  } else {
+    numofRequests = request_ptr->num_of_Shards_ - shard_idx + 1; 
+  }
+
+  // Reserving memory for pending_requests and pending_responses
+  // By default we are setting it to 3.
+  request_ptr->pending_requests_.reserve(numofRequests);
+  request_ptr->pending_responses_.reserve(numofRequests);
+
+  //keeping shard index and pending_requests_index same
+  for (int32_t i = request_ptr->num_of_Shards_-1; i >= shard_idx ; i--) {
+    request_ptr->pending_requests_.emplace_back(*request_ptr, i, i);
```

Comment:
request index and shard index should not be same , as it will lead to crash after the fixing of numofRequests logic as mentioned in above review comment. Fix this

Reply:
Fixed this